File: backend/app/schemas/websocket.py
"""
WebSocket message schemas
These schemas mirror the TypeScript definitions in shared/types/websocket.ts
"""

from pydantic import BaseModel
from typing import Any, Optional, Literal
from datetime import datetime
from enum import Enum


class WebSocketEventType(str, Enum):
    """WebSocket event types"""
    JOB_STARTED = "job_started"
    JOB_PROGRESS = "job_progress"
    JOB_COMPLETED = "job_completed"
    JOB_FAILED = "job_failed"
    COMPANY_UPDATED = "company_updated"
    COMPANY_LOCKED = "company_locked"
    COMPANY_UNLOCKED = "company_unlocked"
    NOTIFICATION = "notification"
    USER_ACTIVITY = "user_activity"
    CAMPAIGN_UPDATED = "campaign_updated"
    EMAIL_DRAFT_UPDATED = "email_draft_updated"


class WebSocketMessage(BaseModel):
    """Base WebSocket message structure"""
    event: WebSocketEventType
    data: Any
    user_id: Optional[str] = None
    timestamp: datetime = datetime.utcnow()

    class Config:
        use_enum_values = True


# Event-specific data schemas
class JobStartedData(BaseModel):
    job_id: str
    job_type: str


class JobProgressData(BaseModel):
    job_id: str
    progress: int  # 0-100
    status_message: Optional[str] = None


class JobCompletedData(BaseModel):
    job_id: str
    job_type: str
    result: Optional[Any] = None
    domain_count: Optional[int] = None


class JobFailedData(BaseModel):
    job_id: str
    job_type: str
    error: str


class CompanyUpdatedData(BaseModel):
    company_id: str
    status: Optional[str] = None
    enriched_at: Optional[datetime] = None
    fields_updated: Optional[list[str]] = None


class CompanyLockedData(BaseModel):
    company_id: str
    locked_by_user_id: str
    locked_by_user_name: Optional[str] = None


class CompanyUnlockedData(BaseModel):
    company_id: str


class NotificationData(BaseModel):
    id: str
    type: Literal["info", "success", "warning", "error"]
    title: str
    message: str
    action_url: Optional[str] = None


class UserActivityData(BaseModel):
    user_id: str
    user_name: str
    action: str
    resource_type: str
    resource_id: str
    description: str


class CampaignUpdatedData(BaseModel):
    campaign_id: str
    status: Optional[str] = None
    generated_count: Optional[int] = None
    sent_count: Optional[int] = None


class EmailDraftUpdatedData(BaseModel):
    draft_id: str
    company_id: str
    status: Optional[str] = None
    gmail_draft_id: Optional[str] = None

File: shared/types/websocket.ts
/**
 * Shared TypeScript definitions for WebSocket events
 * These types ensure type-safe real-time communication
 */

export type WebSocketEventType =
  | 'job_started'
  | 'job_progress'
  | 'job_completed'
  | 'job_failed'
  | 'company_updated'
  | 'company_locked'
  | 'company_unlocked'
  | 'notification'
  | 'user_activity'
  | 'campaign_updated'
  | 'email_draft_updated';

export interface WebSocketMessage<T = any> {
  event: WebSocketEventType;
  data: T;
  user_id?: string;
  timestamp: string;
}

// Event-specific data types
export interface JobStartedData {
  job_id: string;
  job_type: string;
}

export interface JobProgressData {
  job_id: string;
  progress: number;
  status_message?: string;
}

export interface JobCompletedData {
  job_id: string;
  job_type: string;
  result?: any;
  domain_count?: number;
}

export interface JobFailedData {
  job_id: string;
  job_type: string;
  error: string;
}

export interface CompanyUpdatedData {
  company_id: string;
  status?: string;
  enriched_at?: string;
  fields_updated?: string[];
}

export interface CompanyLockedData {
  company_id: string;
  locked_by_user_id: string;
  locked_by_user_name?: string;
}

export interface CompanyUnlockedData {
  company_id: string;
}

export interface NotificationData {
  id: string;
  type: 'info' | 'success' | 'warning' | 'error';
  title: string;
  message: string;
  action_url?: string;
}

export interface UserActivityData {
  user_id: string;
  user_name: string;
  action: string;
  resource_type: string;
  resource_id: string;
  description: string;
}

export interface CampaignUpdatedData {
  campaign_id: string;
  status?: string;
  generated_count?: number;
  sent_count?: number;
}

export interface EmailDraftUpdatedData {
  draft_id: string;
  company_id: string;
  status?: string;
  gmail_draft_id?: string;
}

// Type guards for type-safe event handling
export function isJobProgressData(data: any): data is JobProgressData {
  return 'job_id' in data && 'progress' in data;
}

export function isCompanyUpdatedData(data: any): data is CompanyUpdatedData {
  return 'company_id' in data;
}

export function isNotificationData(data: any): data is NotificationData {
  return 'type' in data && 'title' in data && 'message' in data;
}
